File: backend/app/models/game_models.py
from pydantic import BaseModel, Field
from typing import List, Optional, Dict
from enum import Enum
from datetime import datetime


class RoomType(str, Enum):
    BASIC = "basic"
    STANDARD = "standard"
    DELUXE = "deluxe"
    ROYAL = "royal"


class GuestType(str, Enum):
    PEASANT = "peasant"
    MERCHANT = "merchant"
    NOBLE = "noble"
    ADVENTURER = "adventurer"
    WIZARD = "wizard"
    BANDIT = "bandit"  # Alto guadagno, bassa reputazione
    MONK = "monk"  # Basso guadagno, alta reputazione
    BARD = "bard"  # Medio guadagno, alta reputazione
    DRAGON_DISGUISED = "dragon_disguised"  # Altissimo guadagno, reputazione casuale
    BEGGAR = "beggar"  # Bassissimo guadagno, media reputazione
    PRINCE = "prince"  # Alto guadagno, altissima reputazione
    THIEF = "thief"  # Alto guadagno, reputazione negativa
    SCHOLAR = "scholar"  # Basso guadagno, alta reputazione
    DRUNK = "drunk"  # Medio guadagno, bassa reputazione
    GHOST = "ghost"  # Nessun guadagno, alta reputazione (speciale)


class Room(BaseModel):
    id: str
    room_type: RoomType
    level: int = 1
    occupied: bool = False
    current_guest: Optional[str] = None
    income_rate: float = Field(default=1.0)
    cleanliness: float = Field(default=100.0, ge=0, le=100)


class ItemType(str, Enum):
    FOOD = "food"
    BEVERAGE = "beverage"


class ItemQuality(str, Enum):
    BASIC = "basic"
    GOOD = "good"
    FINE = "fine"
    EXQUISITE = "exquisite"
    LEGENDARY = "legendary"


class IngredientRarity(str, Enum):
    COMMON = "common"
    UNCOMMON = "uncommon"
    RARE = "rare"
    EPIC = "epic"
    LEGENDARY = "legendary"


class Ingredient(BaseModel):
    id: str
    name: str
    rarity: IngredientRarity
    description: str = ""
    # Drop chances from guests
    base_drop_chance: float = 0.1  # 10% base chance


class IngredientRequirement(BaseModel):
    ingredient_id: str
    quantity: int = 1


class TavernItem(BaseModel):
    id: str
    name: str
    item_type: ItemType
    quality: ItemQuality
    cost: float  # Cost to produce/buy
    gold_bonus: float = 0.0  # Extra gold per tick when served
    patience_bonus: float = 0.0  # Patience restored
    reputation_bonus: float = 0.0  # Extra reputation
    satisfaction_bonus: float = 0.0  # Overall satisfaction increase
    description: str = ""


class Recipe(BaseModel):
    id: str
    name: str
    item_id: str  # The item this recipe creates
    unlocked: bool = False
    discovered: bool = False  # Has player discovered this recipe?
    cost_to_unlock: float = 0.0
    ingredients_cost: float = 0.0  # Cost in gold to craft (legacy, still used)
    required_ingredients: List[IngredientRequirement] = Field(default_factory=list)
    # If empty, uses ingredients_cost as gold-only recipe


class Inventory(BaseModel):
    items: Dict[str, int] = Field(default_factory=dict)  # item_id -> quantity


class IngredientInventory(BaseModel):
    ingredients: Dict[str, int] = Field(default_factory=dict)  # ingredient_id -> quantity


class Guest(BaseModel):
    id: str
    name: str
    guest_type: GuestType
    room_id: Optional[str] = None
    patience: float = Field(default=100.0, ge=0, le=100)
    gold_per_tick: float = 1.0
    reputation_bonus: float = 0.1
    check_in_time: Optional[datetime] = None
    stay_duration: int = 10  # ticks
    fed: bool = False  # Has been served food
    served_drink: bool = False  # Has been served beverage
    satisfaction: float = Field(default=50.0, ge=0, le=100)  # Guest satisfaction
    food_served: Optional[str] = None  # ID of food item served
    beverage_served: Optional[str] = None  # ID of beverage served


class Upgrade(BaseModel):
    id: str
    name: str
    description: str
    cost: float
    purchased: bool = False
    effect_type: str  # "income_multiplier", "auto_clean", "guest_capacity", etc.
    effect_value: float


class Resources(BaseModel):
    gold: float = Field(default=100.0)
    reputation: float = Field(default=0.0)
    max_guests: int = Field(default=5)


class InnState(BaseModel):
    resources: Resources = Field(default_factory=Resources)
    rooms: List[Room] = Field(default_factory=list)
    guests: List[Guest] = Field(default_factory=list)
    upgrades: List[Upgrade] = Field(default_factory=list)
    total_income_multiplier: float = 1.0
    auto_clean_enabled: bool = False
    last_update: datetime = Field(default_factory=datetime.utcnow)
    game_speed: float = 1.0  # Ticks per second
    # Tavern system
    tavern_items: List[TavernItem] = Field(default_factory=list)
    recipes: List[Recipe] = Field(default_factory=list)
    inventory: Inventory = Field(default_factory=Inventory)
    tavern_unlocked: bool = False
    # Ingredient system
    available_ingredients: List[Ingredient] = Field(default_factory=list)
    ingredient_inventory: IngredientInventory = Field(default_factory=IngredientInventory)


class GameAction(BaseModel):
    action_type: str
    target_id: Optional[str] = None
    data: Optional[dict] = None

File: frontend/src/app/models/game.models.ts
export enum RoomType {
  BASIC = 'basic',
  STANDARD = 'standard',
  DELUXE = 'deluxe',
  ROYAL = 'royal'
}

export enum GuestType {
  // Common travelers
  PEASANT = 'peasant',
  MERCHANT = 'merchant',
  NOBLE = 'noble',
  ADVENTURER = 'adventurer',

  // Magical types
  WIZARD = 'wizard',
  SORCERER = 'sorcerer',
  WITCH = 'witch',
  ALCHEMIST = 'alchemist',
  NECROMANCER = 'necromancer',

  // Warriors
  KNIGHT = 'knight',
  PALADIN = 'paladin',
  RANGER = 'ranger',
  BARBARIAN = 'barbarian',
  SAMURAI = 'samurai',

  // Rogues/Outlaws
  BANDIT = 'bandit',
  THIEF = 'thief',
  ASSASSIN = 'assassin',
  SMUGGLER = 'smuggler',

  // Religious/Spiritual
  MONK = 'monk',
  PRIEST = 'priest',
  CLERIC = 'cleric',
  DRUID = 'druid',

  // Entertainers
  BARD = 'bard',
  JESTER = 'jester',
  MINSTREL = 'minstrel',
  ACTOR = 'actor',

  // Craftspeople
  BLACKSMITH = 'blacksmith',
  CARPENTER = 'carpenter',
  JEWELER = 'jeweler',
  TAILOR = 'tailor',

  // Scholars
  SCHOLAR = 'scholar',
  SCRIBE = 'scribe',
  LIBRARIAN = 'librarian',
  HISTORIAN = 'historian',

  // Exotic/Rare
  DRAGON_DISGUISED = 'dragon_disguised',
  VAMPIRE = 'vampire',
  WEREWOLF = 'werewolf',
  ELF = 'elf',
  DWARF = 'dwarf',
  HALFLING = 'halfling',
  ORC = 'orc',

  // Special
  BEGGAR = 'beggar',
  PRINCE = 'prince',
  PIRATE = 'pirate',
  SPY = 'spy',
  DRUNK = 'drunk',
  GHOST = 'ghost'
}

export interface Room {
  id: string;
  room_type: RoomType;
  level: number;
  occupied: boolean;
  current_guest: string | null;
  income_rate: number;
  cleanliness: number;
  customers_served: number;
}

export interface RoomTypeTemplate {
  id: string;
  name: string;
  description: string;
  base_cost: number;
  income_multiplier: number;
  emoji: string;
  required_upgrade_id: string | null;
  species_bonuses: { [species: string]: number };
  type_bonuses: { [type: string]: number };
}

export enum ItemType {
  FOOD = 'food',
  BEVERAGE = 'beverage'
}

export enum ItemQuality {
  BASIC = 'basic',
  GOOD = 'good',
  FINE = 'fine',
  EXQUISITE = 'exquisite',
  LEGENDARY = 'legendary'
}

export interface TavernItem {
  id: string;
  name: string;
  item_type: ItemType;
  quality: ItemQuality;
  cost: number;
  gold_bonus: number;
  patience_bonus: number;
  reputation_bonus: number;
  satisfaction_bonus: number;
  description: string;
}

export interface IngredientRequirement {
  ingredient_id: string;
  quantity: number;
}

export interface Recipe {
  id: string;
  name: string;
  item_id: string;
  unlocked: boolean;
  discovered: boolean;
  cost_to_unlock: number;
  ingredients_cost: number;
  required_ingredients: IngredientRequirement[];
  times_crafted: number;
  discovered_at: string | null;
  unlocked_at: string | null;
}

export enum IngredientRarity {
  COMMON = 'common',
  UNCOMMON = 'uncommon',
  RARE = 'rare',
  EPIC = 'epic',
  LEGENDARY = 'legendary'
}

export interface Ingredient {
  id: string;
  name: string;
  rarity: IngredientRarity;
  description: string;
  base_drop_chance: number;
  is_purchasable: boolean;
  market_price: number;
}

export interface ExperimentResult {
  success: boolean;
  message: string;
  discovered_recipe?: {
    recipe_id: string;
    name: string;
    item_name: string;
    unlocked: boolean;
    cost_to_unlock: number;
  };
  consumed_ingredients?: boolean;
  game_state: InnState;
}

export interface Inventory {
  items: { [key: string]: number };
  ingredients: { [key: string]: number };
}

export interface Guest {
  id: string;
  name: string;
  guest_type: GuestType;
  species: string;
  room_id: string | null;
  patience: number;
  gold_per_tick: number;
  reputation_bonus: number;
  check_in_time: string | null;
  stay_duration: number;
  fed: boolean;
  served_drink: boolean;
  satisfaction: number;
  food_served: string | null;
  beverage_served: string | null;
}

export interface Upgrade {
  id: string;
  name: string;
  description: string;
  cost: number;
  purchased: boolean;
  effect_type: string;
  effect_value: number;
  operational_cost_per_tick?: number;
  // Premium upgrade fields
  is_premium?: boolean;
  premium_price_cents?: number;
  premium_price_usd?: number;
  duration_seconds?: number;
  is_consumable?: boolean;
}

export interface Resources {
  gold: number;
  reputation: number;
  max_guests: number;
}

export interface Location {
  map: string;
  x: number;
  y: number;
}

export interface OfflineProgress {
  earnings: number;
  hours: number;
}

export interface Achievement {
  id: string;
  name: string;
  description: string;
  requirement_type: string;
  requirement_value: number;
  requirement_metadata: any;
  icon: string;
  reward_upgrade_id: string | null;
  reward_upgrade_name: string | null;
}

export interface PlayerAchievement {
  achievement: Achievement;
  progress: number;
  earned_at: string;
  is_completed: boolean;
}

export interface AchievementWithProgress {
  id: string;
  name: string;
  description: string;
  requirement_type: string;
  requirement_value: number;
  requirement_metadata: any;
  icon: string;
  reward_upgrade_id: string | null;
  reward_upgrade_name: string | null;
  progress: number;
  is_completed: boolean;
  earned_at: string | null;
}

export interface InnState {
  resources: Resources;
  rooms: Room[];
  guests: Guest[];
  upgrades: Upgrade[];
  total_income_multiplier: number;
  auto_clean_enabled: boolean;
  last_update: string;
  game_speed: number;
  tavern_items: TavernItem[];
  recipes: Recipe[];
  inventory: Inventory;
  tavern_unlocked: boolean;
  available_ingredients: Ingredient[];
  room_types: RoomTypeTemplate[];
  location: Location;
  offline_progress: OfflineProgress;
  max_offline_hours: number;
  achievements: PlayerAchievement[];
  all_achievements: AchievementWithProgress[];
  premium_upgrades: Upgrade[];
}
